pkg/power/platform/source/redfish.go

```diff
@@ -0,0 +1,286 @@
+				klog.V(5).Infof("Initialized redfish credential")
+				probeInterval := config.GetRedfishProbeIntervalInSeconds()
+				interval := time.Duration(probeInterval) * time.Second
+				redfish := &RedFishClient{
+					accessInfo:    RedfishAccessInfo{Username: userName, Password: password, Host: host},
+					systems:       []*RedfishSystemPowerResult{},
+					probeInterval: interval,
+					mutex:         sync.Mutex{},
+				}
+				return redfish
+			}
+		} else {
+			klog.V(1).Infof("%s", fmt.Sprintf("failed to get node credential: %v", err))
+			return nil
+		}
+	}
+	return nil
+}
+
+func (rf *RedFishClient) IsSystemCollectionSupported() bool {
+	system, err := getRedfishSystem(rf.accessInfo)
+
+	if err != nil {
+		klog.Infof("failed to get redfish system info: %v\n", err)
+		return false
+	}
+
+	intervalInMin := 0
+	// iterate each "Members" in the system and get the power info
+	for _, member := range system.Members {
+		// split the OdataID by delimiter "/" and get the system ID
+		split := strings.Split(member.OdataID, "/")
+		if len(split) < 2 {
+			continue
+		}
+		id := split[len(split)-1]
+		res := RedfishSystemPowerResult{}
+		power, err := getRedfishPower(rf.accessInfo, id)
+		if err == nil && len(power.PowerControl) > 0 {
+			res.system = id
+			res.consumedWatts = power.PowerControl[0].PowerConsumedWatts
+			res.timestamp = time.Now()
+			rf.systems = append(rf.systems, &res)
+			klog.V(5).Infof("power info: %+v\n", power)
+			if power.PowerControl[0].PowerMetrics.IntervalInMin > intervalInMin {
+				intervalInMin = power.PowerControl[0].PowerMetrics.IntervalInMin
+			}
+		} else {
+			klog.V(5).Infof("failed to get power info: %v\n", err)
+		}
+	}
+
+	// set a timer to check the power info every probeInterval seconds
+	rf.ticker = time.NewTicker(rf.probeInterval)
+	go func() {
+		for {
+			<-rf.ticker.C
+			for _, system := range rf.systems {
+				power, err := getRedfishPower(rf.accessInfo, system.system)
+				if err == nil && len(power.PowerControl) > 0 {
+					// mutex
+					rf.mutex.Lock()
+					klog.V(5).Infof("power info: %+v\n", power)
+					system.consumedWatts = power.PowerControl[0].PowerConsumedWatts
+					system.timestamp = time.Now()
+					rf.mutex.Unlock()
+				} else {
+					klog.V(5).Infof("failed to get power info: %v\n", err)
+				}
+			}
+		}
+	}()
+	return rf.systems != nil && len(rf.systems) > 0
+}
+
+// GetEnergyFromPlatform returns the power consumption in Watt
+func (rf *RedFishClient) GetEnergyFromPlatform() (map[string]float64, error) {
+	if rf.systems != nil {
+		power := make(map[string]float64)
+		for _, system := range rf.systems {
+			rf.mutex.Lock()
+			now := time.Now()
+			// calculate the elapsed time since the last power query in seconds
+			elapsed := now.Sub(system.timestamp).Seconds()
+			system.timestamp = time.Now()
+			klog.V(5).Infof("power info: %+v\n", system)
+			power[system.system] = float64(system.consumedWatts * 1000 * int(elapsed)) // convert to mW
```

Comment:
```suggestion
			power[system.system] = float64(system.consumedWatts*1000) * elapsed // convert to mW
```

Reply:
oh, i see. This function is get energy (mJ), instead of power. The comment needs to be fixed but the formula is right. See the ACPI function [here](https://github.com/sustainable-computing-io/kepler/blob/main/pkg/power/platform/source/acpi.go#L166)

Reply:
I know the formula is right. I mean cast the elapsed time from float to int might lose energy of around one second if it was 2.999... .